jetty-client/src/main/java/org/eclipse/jetty/client/HttpReceiver.java

```diff
@@ -307,45 +365,56 @@ protected boolean responseHeaders(HttpExchange exchange)
      */
     protected boolean responseContent(HttpExchange exchange, ByteBuffer buffer, Callback callback)
     {
-        out:
         while (true)
         {
             ResponseState current = responseState.get();
-            switch (current)
+            if (current == ResponseState.HEADERS || current == ResponseState.CONTENT)
             {
-                case HEADERS:
-                case CONTENT:
-                {
-                    if (updateResponseState(current, ResponseState.TRANSIENT))
-                        break out;
+                if (updateResponseState(current, ResponseState.TRANSIENT))
                     break;
-                }
-                default:
-                {
-                    callback.failed(new IllegalStateException("Invalid response state " + current));
-                    return false;
-                }
+            }
+            else
+            {
+                callback.failed(new IllegalStateException("Invalid response state " + current));
+                return false;
             }
         }
 
+        if (demand() <= 0)
+        {
+            callback.failed(new IllegalStateException("No demand for response content"));
+            return false;
+        }
+
         HttpResponse response = exchange.getResponse();
         if (LOG.isDebugEnabled())
             LOG.debug("Response content {}{}{}", response, System.lineSeparator(), BufferUtil.toDetailString(buffer));
 
-        ResponseNotifier notifier = getHttpDestination().getResponseNotifier();
-
-        ContentDecoder decoder = this.decoder;
-        if (decoder == null)
+        if (contentListeners.isEmpty())
         {
-            notifier.notifyContent(response, buffer, callback, contentListeners);
+            callback.succeeded();
         }
         else
         {
-            new Decoder(notifier, response, decoder, buffer, callback).iterate();
+            Decoder decoder = this.decoder;
+            if (decoder == null)
+            {
+                contentListeners.notifyContent(response, buffer, callback);
+            }
+            else
+            {
+                if (!decoder.decode(buffer, callback))
+                    return false;
+            }
         }
 
         if (updateResponseState(ResponseState.TRANSIENT, ResponseState.CONTENT))
-            return true;
+        {
+            boolean hasDemand = hasDemandOrStall();
```

Comment:
Is this the only place the state can leave `TRANSIENT`?   Ie is the terminateResponse below an error state?  If so would be good to indicate that.  If not, would be good to comment who we are racing with.